Update tandem duplication annotations

SHuang-Broad opened the pull request:
This PR:

* fixes a problem observed for a complex event with tandem duplicated sequence and insertion and deletion, which causes the discovery phase of SV pipeline to throw exceptions;
* the fix works by changing how tandem duplications are annotated:
    * duplicated sequence is no longer provided, instead the corresponding duplication unit reference span is provided in `DUP_REPET_UNIT_REF_SPAN`,
    * the duplicated units on the assembled contigs' CIGAR when aligned to reference is provided in `DUP_ASM_CTG_CIGARS`;
* adds annotation `DUP_ANNOT_FROM_OPT` when tandem repeat annotations were generated by a simple approximation procedure, which should be viewed with care;
* logs the total number of variants and different types
* updated and added tests to reflect these changes

The PR was tested to be runnable based on output from scanning the CHM-mix bam with PR #2444, which discovered the exception.
The number of variants discovered are:

For CHM-mix
```
20:43:36.213 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 204 INVs
20:43:36.213 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 2775 DELs
20:43:36.213 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 954 DUPs
20:43:36.213 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 977 INSs
```
And for NA12878_PCR-_30X
```
22:14:15.653 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - Discovered 4686 variants.
22:14:15.660 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 228 INV's
22:14:15.660 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 2719 DEL's
22:14:15.660 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 835 DUP's
22:14:15.660 INFO  DiscoverStructuralVariantsFromAlignedContigsSAMSpark - 904 INS's
```
@cwhelan Could you please to review?
@tedsharpe feel free to poke around and test run it.


__UPDATE__:
This is to be merged after PR #2444.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
```diff
@@ -31,9 +31,13 @@
     public static final String INV33 = "INV33";
     public static final String INV55 = "INV55";
-    public static final String DUPLICATED_SEQUENCE = "DUPLICATED_SEQUENCE";
+//    public static final String DUPLICATED_SEQUENCE = "DUPLICATED_SEQUENCE";
```
remove commented line

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
```diff
@@ -31,9 +31,13 @@
-    public static final String DUPLICATED_SEQUENCE = "DUPLICATED_SEQUENCE";
+//    public static final String DUPLICATED_SEQUENCE = "DUPLICATED_SEQUENCE";
+    public static final String DUP_REPET_UNIT_REF_SPAN = "DUP_REPET_UNIT_REF_SPAN";
+    public static final String DUP_ASM_CTG_CIGARS = "DUP_ASM_CTG_CIGARS";
```
Do you think `DUP_SEQ_CIGARS` might capture this and be easier to understand?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
```diff
@@ -31,9 +31,13 @@
+    public static final String DUP_ASM_CTG_CIGARS = "DUP_ASM_CTG_CIGARS";
     public static final String DUPLICATION_NUMBERS = "DUP_NUM";
 
+    public static final String DUP_ANNOT_FROM_OPT = "DUP_ANNOT_FROM_OPT";
```
It looks like we're always setting this, do we really need it? What if you just put an extra caveat in the header line descriptions? If you want to keep it, I might change the name to something more general, like "DUP_ANNOTATIONS_ESTIMATED" or "DUP_ANNOTATIONS_IMPRECISE".

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
```diff
@@ -56,8 +60,12 @@
         vcfHeaderLines.put(INV55, new VCFInfoHeaderLine(INV55, 0, VCFHeaderLineType.Flag, "Whether the event represents a 5' to 3' inversion"));
 
-        vcfHeaderLines.put(DUPLICATED_SEQUENCE, new VCFInfoHeaderLine(DUPLICATED_SEQUENCE, 1, VCFHeaderLineType.String, "Duplicated sequence on the reference or on the alternate allele"));
+//        vcfHeaderLines.put(DUPLICATED_SEQUENCE, new VCFInfoHeaderLine(DUPLICATED_SEQUENCE, 1, VCFHeaderLineType.String, "Duplicated sequence on the reference or on the alternate allele"));
```
remove commented line

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/NovelAdjacencyReferenceLocations.java:
```diff
@@ -157,260 +60,20 @@ static EndConnectionType determineEndConnectionType(final ChimericAlignment chim
         final int leftBreakpointCoord, rightBreakpointCoord;
-        if (complication.dupSeqRepeatNumOnRef !=0 && complication.dupSeqRepeatNumOnCtg !=0) { // todo : development artifact-- assuming tandem duplication is not co-existing with inversion
-            final int dupLen = complication.dupSeqForwardStrandRep.length();
+        if (!complication.dupSeqRepeatUnitRefSpan.equals(BreakpointComplications.DUPSEQ_REPEAT_UNIT_NA_VALUE)) { // todo : development artifact-- assuming tandem duplication is not co-existing with inversion
```
As I said earlier, it'd be nice to hide references to `BreakpointComplications.DUPSEQ_REPEAT_UNIT_NA_VALUE` behind a more straightforwardly named method.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusCall.java:
```diff
@@ -91,7 +92,7 @@ static SvType getType(final NovelAdjacencyReferenceLocations novelAdjacencyRefer
         if (endConnectionType == FIVE_TO_THREE) { // no strand switch happening, so no inversion
             if (start==end) { // something is inserted
-                final boolean hasNoDupSeq = novelAdjacencyReferenceLocations.complication.dupSeqForwardStrandRep.isEmpty();
+                final boolean hasNoDupSeq = novelAdjacencyReferenceLocations.complication.dupSeqRepeatUnitRefSpan.equals(BreakpointComplications.DUPSEQ_REPEAT_UNIT_NA_VALUE);
```
Hide `BreakpointComplications.DUPSEQ_REPEAT_UNIT_NA_VALUE`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusCall.java:
```diff
@@ -153,9 +154,16 @@ static SvType getType(final NovelAdjacencyReferenceLocations novelAdjacencyRefer
-        if (!novelAdjacencyReferenceLocations.complication.dupSeqForwardStrandRep.isEmpty()) {
-            attributeMap.put(GATKSVVCFHeaderLines.DUPLICATED_SEQUENCE, novelAdjacencyReferenceLocations.complication.dupSeqForwardStrandRep);
+        if (!(novelAdjacencyReferenceLocations.complication.dupSeqRepeatUnitRefSpan.equals(BreakpointComplications.DUPSEQ_REPEAT_UNIT_NA_VALUE))) {
+//            attributeMap.put(GATKSVVCFHeaderLines.DUPLICATED_SEQUENCE, novelAdjacencyReferenceLocations.complication.dupSeqForwardStrandRep);
```
remove commented line

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SvType.java:
```diff
@@ -82,12 +82,12 @@ public String toString() {
                     Allele.create(SVConstants.CallingStepConstants.VCF_ALT_ALLELE_STRING_DEL),
                     -(novelAdjacencyReferenceLocations.leftJustifiedRightRefLoc.getStart() - novelAdjacencyReferenceLocations.leftJustifiedLeftRefLoc.getEnd()),
-                    novelAdjacencyReferenceLocations.complication.dupSeqForwardStrandRep.isEmpty() ? Collections.EMPTY_MAP : Collections.singletonMap(SVConstants.CallingStepConstants.TANDUP_CONTRACTION_STRING, ""));
+                    novelAdjacencyReferenceLocations.complication.dupSeqRepeatUnitRefSpan.equals(BreakpointComplications.DUPSEQ_REPEAT_UNIT_NA_VALUE) ? Collections.EMPTY_MAP : Collections.singletonMap(SVConstants.CallingStepConstants.TANDUP_CONTRACTION_STRING, ""));
```
hide

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/NovelAdjacencyReferenceLocationsUnitTest.java:
```diff
@@ -376,7 +383,8 @@ public void testGetBreakpoints_tandemDuplication_complex() throws IOException {
         Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.homologyForwardStrandRep, pseudoHomology)<=2);
         Assert.assertTrue(breakpoints.complication.insertedSequenceForwardStrandRep.isEmpty());
-        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
+//        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
```
remove

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/NovelAdjacencyReferenceLocationsUnitTest.java:
```diff
@@ -386,7 +394,8 @@ public void testGetBreakpoints_tandemDuplication_complex() throws IOException {
         Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.homologyForwardStrandRep, pseudoHomology)<=2);
         Assert.assertTrue(breakpoints.complication.insertedSequenceForwardStrandRep.isEmpty());
-        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
+//        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
```
remove

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/NovelAdjacencyReferenceLocationsUnitTest.java:
```diff
@@ -397,8 +406,8 @@ public void testGetBreakpoints_tandemDuplication_complex() throws IOException {
         Assert.assertTrue(breakpoints.complication.insertedSequenceForwardStrandRep.isEmpty());
-        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
-        Assert.assertEquals(breakpoints.complication.dupSeqRepeatNumOnRef, 2);
+//        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
```
remove.. and the other commented lines in this class

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/NovelAdjacencyReferenceLocationsUnitTest.java:
```diff
@@ -407,7 +416,8 @@ public void testGetBreakpoints_tandemDuplication_complex() throws IOException {
         Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.homologyForwardStrandRep, firstRepeat+pseudoHomology)<=2);
         Assert.assertTrue(breakpoints.complication.insertedSequenceForwardStrandRep.isEmpty());
-        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
+//        Assert.assertTrue(StringUtils.getLevenshteinDistance(breakpoints.complication.dupSeqForwardStrandRep, firstRepeat)<=2);
```
remove

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/NovelAdjacencyReferenceLocationsUnitTest.java:
```diff
@@ -449,30 +462,66 @@ public void testGetBreakpoints_tandemDuplication_complex() throws IOException {
-//    }
+    // TODO: 4/3/17 enable when transolocations can be handled
+    @Test(enabled = false)
+    public void testGetAssembledBreakpointsFromAlignmentRegions() throws Exception {
```
Does this test the breaking use case? Could you add a comment or two to the test describing what it's testing?

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusCallUnitTest.java:
```diff
@@ -258,49 +258,49 @@ public void testIntegrative() throws IOException {
 
         seeIfItWorks_integrative(testData, Stream.concat( commonAttributes.stream(),
-                Sets.newHashSet(SVConstants.CallingStepConstants.TANDUP_CONTRACTION_STRING, GATKSVVCFHeaderLines.DUPLICATED_SEQUENCE, GATKSVVCFHeaderLines.DUPLICATION_NUMBERS, GATKSVVCFHeaderLines.HOMOLOGY, GATKSVVCFHeaderLines.HOMOLOGY_LENGTH).stream())
+                Sets.newHashSet(SVConstants.CallingStepConstants.TANDUP_CONTRACTION_STRING, /*GATKSVVCFHeaderLines.DUPLICATED_SEQUENCE,*/ GATKSVVCFHeaderLines.DUP_REPET_UNIT_REF_SPAN, GATKSVVCFHeaderLines.DUPLICATION_NUMBERS, GATKSVVCFHeaderLines.HOMOLOGY, GATKSVVCFHeaderLines.HOMOLOGY_LENGTH).stream())
```
Remove these commented arguments

SHuang-Broad commented:
@cwhelan I've addressed the comments in the separate commits. The commit messages decribes what was done in each push.. Please take another look. Thanks!

The tests are failing because of travis. They run successfully when manually rebooted but the status here don't get updated...

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
What if you called this inner class `TandemRepeatStructure` or something? Having an inner class that ends in the same word (Complication) as the outer class made me think it would be a subclass at first reading, but this is really storing extra stuff.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVCFWriter.java:
```diff
@@ -33,15 +34,27 @@
+
+        sortedVariantsList.stream()
+                .collect(Collectors.groupingBy(vc -> (String)vc.getAttribute(GATKSVVCFHeaderLines.SVTYPE)))
+                .entrySet().forEach(pair -> logger.info(pair.getValue().size() + " "+ pair.getKey() + "'s"));
```
I'm not sure, but it might be more efficient to use a counting collector here. I think it would look like this (not tested):

```
sortedVariantsList.stream()
   .collect(Collectors.groupingBy(vc -> (String)vc.getAttribute(GATKSVVCFHeaderLines.SVTYPE)), Collectors.counting())
  .entrySet().forEach(pair -> logger.info(pair.getKey() + ": " + pair.getValue"));

```

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
I want all `BreakpointComplications` objects to be effectively immutable. This was done previously by making all fields final (I know not immutable, but enough approximation). With the new way of initializing the fields, they cannot be `final` so I provided getters but no setters. Therefore I need such a ctor, to provided opportunity when the default initialization is not enought, which is the case in the test with inversion and homology around breakpoints.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
yep.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
moved to test with new default ctor

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
all removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done, don't know why...

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
the default is to be in sync with `hasDuplicationAnnotation`: null when that is false

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
OK. I might put in a comment that says the constructor is only for test purposes, then.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
done

tedsharpe commented:
@VisibleForTesting?

SHuang-Broad commented:
@cwhelan update per comments. Please take another look.
The "bug" of ASSEMBLY_IDS bug would be automatically corrected once Ted' PR #2444 is in.

Thanks!

cwhelan commented:
Looks good to me now!

SHuang-Broad commented:
@cwhelan . Thanks! Will merge after #2444 is in, so that PR doesn't have to resolve conflicts again.
